fedimint-core/src/task.rs

```diff
@@ -379,16 +398,34 @@ impl TaskHandle {
     ///
     /// Tasks can use `select` on the return value to handle shutdown
     /// signal during otherwise blocking operation.
-    pub async fn make_shutdown_rx(&self) -> oneshot::Receiver<()> {
-        let (shutdown_tx, shutdown_rx) = oneshot::channel();
-        self.on_shutdown(Box::new(|| {
-            Box::pin(async {
-                let _ = shutdown_tx.send(());
-            })
-        }))
-        .await;
+    pub async fn make_shutdown_rx(&self) -> TaskShutdownToken {
+        TaskShutdownToken(self.inner.on_shutdown_rx.clone())
+    }
+}
+
+pub struct TaskShutdownToken(watch::Receiver<bool>);
+
+impl Future for TaskShutdownToken {
+    type Output = ();
+
+    fn poll(
+        mut self: Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Self::Output> {
+        let inner = async move {
+            loop {
+                if self.0.changed().await.is_err() {
+                    // if disconnected, shutdown
+                    break;
+                }
+
+                if *self.0.borrow() {
+                    break;
+                }
+            }
+        };
 
-        shutdown_rx
+        Box::pin(inner).as_mut().poll(cx)
```

Comment:
We can save one heap allocation here:

```suggestion
        futures::pin_mut!(inner);
        inner.poll(cx)
```

Reply:
The whole thing is buggy. I can't be recreating that inner future on every call. Do you think there's still a way to avoid Box?